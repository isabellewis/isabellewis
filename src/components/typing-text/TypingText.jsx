import React, { Component } from 'react';
import { Heading } from '../../elements';

import {
  HoverTyping,
} from './TypingText.style';

export class TypingText extends Component {
	constructor(props) {
    super(props);

		this.state = {
			textIndex: 0,
			textDisplayed: 'Isabel',
			animating: false,
			deleting: false,
			writing: false,
		};

		this.startAnimation = this.startAnimation.bind(this);
	}

	textChange() {
		const {
			textIndex,
			textDisplayed,
			deleting,
			writing,
		} = this.state;

		const {
			textValues
		} = this.props;

		if (deleting && textDisplayed.length > 0) {
			this.setState(prevState => ({
				textDisplayed: prevState.textDisplayed.slice(0, -1)
			}));
		} else if (deleting && textDisplayed.length === 0) {
			const nextIndex = textIndex < textValues.length - 1 ? textIndex + 1 : 0;
			this.setState(prevState => ({
				deleting: false,
				writing: true,
				textIndex: nextIndex,
				textDisplayed: prevState.textDisplayed + textValues[nextIndex][0],
			}));
		} else if (writing && textDisplayed.length < textValues[textIndex].length) {
			this.setState(prevState => ({
				textDisplayed: prevState.textDisplayed + textValues[prevState.textIndex][prevState.textDisplayed.length],
			}));
		} else {
			this.endAnimation();
		}
	}

	endAnimation() {
		this.setState({
			animating: false,
			deleting: false,
			writing: false,
		});
		clearInterval(this.interval);
	}

	startAnimation() {
		if (!this.state.animating) {
			this.setState({
				animating: true,
				deleting: true,
			});
			this.interval = setInterval(() => this.textChange(), 500);
		}
	}

	render() {
		return (
      <HoverTyping onMouseOver={this.startAnimation}>
        <Heading title={this.state.textDisplayed} />
      </HoverTyping>
		);
	}
}
