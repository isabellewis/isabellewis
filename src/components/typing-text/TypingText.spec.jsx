import React from 'react';
import { shallow } from 'enzyme';
import { TypingText } from './TypingText';

describe('views/TypingText', () => {
	let component;

	beforeEach(() => {
		component = shallow(<TypingText textValues={['Hello', 'TestValue1', 'TestValue2']} />);
	});

	it('should match the snapshot', () => {
		expect(component).toMatchSnapshot();
	});

	it('should delete the next letter', () => {
		component.instance().setState({
			textIndex: 0,
			textDisplayed: 'Hel',
			deleting: true,
			writing: false
		});

		component.instance().textChange();
		expect(component.instance().state.textIndex).toEqual(0);
		expect(component.instance().state.textDisplayed).toEqual('He');
		expect(component.instance().state.deleting).toEqual(true);
		expect(component.instance().state.writing).toEqual(false);
	});

	it('should change the state to write when the word has been deleted', () => {
		component.instance().setState({
			textIndex: 0,
			textDisplayed: '',
			deleting: true,
			writing: false
		});

		component.instance().textChange();
		expect(component.instance().state.textIndex).toEqual(1);
		expect(component.instance().state.textDisplayed).toEqual('T');
		expect(component.instance().state.deleting).toEqual(false);
		expect(component.instance().state.writing).toEqual(true);
	});

	it('should handle the nextIndex when last word in array has been animated', () => {
		component.instance().setState({
			textIndex: 2,
			textDisplayed: '',
			deleting: true,
			writing: false
		});

		component.instance().textChange();
		expect(component.instance().state.textIndex).toEqual(0);
		expect(component.instance().state.textDisplayed).toEqual('H');
		expect(component.instance().state.deleting).toEqual(false);
		expect(component.instance().state.writing).toEqual(true);
	});

	it('should write the next letter when in the writing state', () => {
		component.instance().setState({
			textIndex: 0,
			textDisplayed: 'Hel',
			deleting: false,
			writing: true
		});

		component.instance().textChange();
		expect(component.instance().state.textIndex).toEqual(0);
		expect(component.instance().state.textDisplayed).toEqual('Hell');
		expect(component.instance().state.deleting).toEqual(false);
		expect(component.instance().state.writing).toEqual(true);
	});

	it('should call endAnimation when finished typing', () => {
		component.instance().setState({
			textIndex: 0,
			textDisplayed: 'Hello',
			deleting: false,
			writing: true
		});

		component.instance().textChange();
		expect(component.instance().state.textIndex).toEqual(0);
		expect(component.instance().state.animating).toEqual(false);
		expect(component.instance().state.textDisplayed).toEqual('Hello');
		expect(component.instance().state.deleting).toEqual(false);
		expect(component.instance().state.writing).toEqual(false);
	});

	it('should set the correct state when startAnimation is called', () => {
		component.instance().setState({
			animating: false,
			textIndex: 0,
			textDisplayed: 'Hello',
			deleting: false,
			writing: false
		});

		component.instance().startAnimation();
		expect(component.instance().state.textIndex).toEqual(0);
		expect(component.instance().state.animating).toEqual(true);
		expect(component.instance().state.textDisplayed).toEqual('Hello');
		expect(component.instance().state.deleting).toEqual(true);
		expect(component.instance().state.writing).toEqual(false);
	});

	it('should not change state if animating is true', () => {
		component.instance().setState({
			animating: true,
			textIndex: 0,
			textDisplayed: 'Hello',
			deleting: false,
			writing: false
		});

		component.instance().startAnimation();
		expect(component.instance().state.textIndex).toEqual(0);
		expect(component.instance().state.animating).toEqual(true);
		expect(component.instance().state.textDisplayed).toEqual('Hello');
		expect(component.instance().state.deleting).toEqual(false);
		expect(component.instance().state.writing).toEqual(false);
	});
});
