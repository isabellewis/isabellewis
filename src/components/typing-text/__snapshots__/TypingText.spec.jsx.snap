// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/TypingText should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TypingText
    textValues={
      Array [
        "Hello",
        "TestValue1",
        "TestValue2",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Heading
        title="Isabel"
      />,
      "onMouseOver": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "title": "Isabel",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Heading
          title="Isabel"
        />,
        "onMouseOver": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "title": "Isabel",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
